### ISSUE TITLE
Memory leak when using Metal delegate

### ISSUE THREAD

**Issue by zhihuiloke** [tensorflow/tensorflow#57718]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Performance

### Source

source

### Tensorflow Version

TFLite 2.9.1

### Custom Code

Yes

### OS Platform and Distribution

iOS 15

### Mobile device

iPhone X

### Python version

N.A.

### Bazel version

5.0.0

### GCC/Compiler version

Apple clang 13.1.6

### CUDA/cuDNN version

N.A.

### GPU model and memory

iPhone X

### Current Behaviour?

```shell
There seems to be a memory leak while using Metal delegate. It doesn't go away no matter how I try to clean up (be it reusing the delegate or not). I have tried to use without the Metal delegate and the leak doesn't happen.
```

### Standalone code to reproduce the issue

```shell
TfLiteDelegate *CreateGPUDelegate(MetalDelegateWaitType waitType = MetalDelegateWaitType::PASSIVE, bool allowPrecisionLoss = false, bool enableQuantization = true)
{
    TFLGpuDelegateOptions options;
    switch(waitType)
    {
    case PASSIVE:
      options.wait_type = TFLGpuDelegateWaitType::TFLGpuDelegateWaitTypePassive;
      break;
    case ACTIVE:
      options.wait_type = TFLGpuDelegateWaitType::TFLGpuDelegateWaitTypeActive;
      break;
    case DO_NOT_WAIT:
      options.wait_type = TFLGpuDelegateWaitType::TFLGpuDelegateWaitTypeDoNotWait;
      break;
    case AGGRESIVE:
      options.wait_type = TFLGpuDelegateWaitType::TFLGpuDelegateWaitTypeAggressive;
      break;
    }
    options.enable_quantization = enableQuantization;
    options.allow_precision_loss = allowPrecisionLoss;
    return TFLGpuDelegateCreate(&options);
}
void DeleteGPUDelegate(TfLiteDelegate *delegate)
{
  TFLGpuDelegateDelete(delegate);
}

// Somewhere in main
  std::thread([this]() {
    std::unique_ptr<tflite::FlatBufferModel> model;
    model = tflite::FlatBufferModel::BuildFromFile(path.c_str());
    while (true)
    {
      std::unique_ptr<tflite::Interpreter> interpreter;
      tflite::ops::builtin::BuiltinOpResolver resolver;
      const TfLiteStatus isInterpreterOk = tflite::InterpreterBuilder(*model, resolver)(&interpreter);
      TfLiteDelegate *delegate = CreateGPUDelegate();
      interpreter->ModifyGraphWithDelegate(delegate);
      interpreter->Invoke();
      interpreter = nullptr;
      DeleteGPUDelegate(delegate);
      std::this_thread::sleep_for(std::chrono::seconds(1));
    }
  }).detach();
```

### Relevant log output

_No response_</details>

**Comment by zhihuiloke** [comment]
Some more info:
- This is built on C++/CMake, with the exception of TensorflowLiteCMetal_framework via Bazel
- I have also tried to reuse the delegate, but it didn't help with the leak

**Comment by mohantym** [comment]
Hi @zhihuiloke !
Probably duplicate to #56287.  Could you check the workaround in this comment and mark this as duplicate.
Thank you!

**Comment by zhihuiloke** [comment]
Yes, this is definitely related to #56287. Sorry for the duplicate. However, I don't see the workaround in that thread, it seems unresolved.

**Comment by zhihuiloke** [comment]
If it is relating to auto release pools, we can't do that either, because we are writing everything in native C++.

**Comment by mohantym** [comment]
Ok @zhihuiloke !
Thanks for the update. Closing you close this issue as duplicate and comment there.

**Comment by zhihuiloke** [comment]
Closing as duplicate

**Comment by zhihuiloke** [comment]
So to have to reopen this. I found out that my problem was slightly different from #56287. From my own research it seems that the leak is coming from ModifyGraphWithDelegate, instead of Invoke.
Attached is a screenshot of instruments, and a 5MB growth per call.
![image](https://user-images.githubusercontent.com/48172598/191225054-cfc30bd2-cc8d-4b42-97c5-131ffdbde9d3.png)

**Comment by mohantym** [comment]
Ok @zhihuiloke !
Thanks for the clarification.

@sachinprasadhs !
Could you look at this issue.

Thank you!

**Comment by mengran1234** [comment]
do you solved problem?
From my own investigation it seems that the leak is coming from ModifyGraphWithDelegate, instead of Invoke.
Attached is a screenshot of instruments, and a 5MB growth per call.

**Comment by mengran1234** [comment]
I alse find this memory leak

**Comment by mengran1234** [comment]
> Closing as duplicate

11

**Comment by mengran1234** [comment]
Hi thank you

**Comment by mengran1234** [comment]
can  you solved problem? @zhihuiloke ，thanks

**Comment by zhihuiloke** [comment]
No. I have not found a solution for it yet.

**Comment by mengran1234** [comment]
@zhihuiloke ，please I think iphone （OC code）has ARC  auto memory manage.Theoretically it has not this leak。why NSString * do not release

**Comment by zhihuiloke** [comment]
I am not too sure what I did, but the leak didn't happen after I changed some stuff.
Also, there were some mistakes in the example code I posted above. The interpreter and ModifyDelegate should have been outside the loop.
I fixed it by:
- Removing the loop that keeps Invoking. Instead, now I spin a thread to run the Invoke.
My suspicion is that the loop somehow prevented the ARC from working, but I could be wrong.

Regardless, I will be closing the issue, since it doesn't really happen to me anymore.
